src/wasm/extract.cpp

```diff
@@ -64,8 +70,12 @@ bool file_output::write(char* data, size_t n) {
 
   position_ += n;
   total_written_ += n;
-
-  if (is_complete()) {
+  if (extractor::get().get_settings().debug_messages_enabled) {
+    printf("write: position=%llu, total_written=%llu, fsize=%llu, size2=%zu, n=%zu\n", position_,
+           total_written_, file_->entry().size, size2, n);
+    std::cout << "write: is_complete():" << std::endl;
+  }
+  if (file_->entry().size != 0 && is_complete() || size2 == total_written_) {
```

Comment:
We will have different combinations of situations when we don't call:
`zs_entryend(zip_, zip_entry_, 0);`

I assume some of them might be expected, others not, I would recommend to add proper logs, especially warnings/errors in unexpected cases.

Reply:
The original is_complete() only checks whether file_->entry().size == total_written_, while this parameter is not always set for an unknown, probably installer-related reason. Thus an additional parameter of size2 with the actual file size that comes from another file descriptor was added to fix this issue and ensure the file is written properly.
This might've not been an issue with the original code, as it was only writing directly to the filesystem, and closing the files was always done at least by the OS when exiting the process, or somehow done via C++ interface.

To sum up: this whole _if_ statement performs a single check - to see if the file was written completely. Maybe it could be rewritten to be a bit more clear, but it's not simply a matter of rewriting is_complete(), as it's not aware of the size2, and it's not a part of the object it's being called on.